### DISCUSSION TITLE
Is Tiered Storage currently planned?

### DISCUSSION THREAD

**Discussion by shafqatevo** [nats-io/nats-server#3871] [General]
Hi,

Couldn't find Tiered Storage in the roadmap (https://nats.io/about/#roadmap). Can anyone pls share what's the current plan for this much-needed feature? What timeframe can it be expected?

**Comment by derekcollison** [comment]
We have it planned but no schedule yet on when we will do this work as of yet.

**Reply by shafqatevo** [reply]
Thanks Derek!

This is a major feature many will be eager to get.

**Reply by Material-Scientist** [reply]
Hi, is there an update on this?

Kafka has recently added tired storage support with the upgrade to KRAFT. I'm not sure how others feel about it, but for me it's the only thing that keeps me from dropping Kafka in favor of NATS Jetstream.

**Reply by derekcollison** [reply]
Can you tell us more about what you would value the most from tiered storage?

**Reply by Material-Scientist** [reply]
So, the (experimental) Kafka tiered storage class (`org.apache.kafka.server.log.remote.storage.LocalTieredStorage`) just uses a local directory to [move segments once they are rolled](https://kafka.apache.org/documentation/#tiered_storage_config_ex).

In my case, that directory is a mounted [SeaweedFS ](https://github.com/seaweedfs/seaweedfs) drive (append-only store), which first stores the data on a fast drive (e.g. SSD), and eventually moves it to slower tiers (e.g. HDD, S3).

Both Kafka & SeaweedFS cache the data back to the local drive as soon as a consumer requires old segments.

Since storage is often the most expensive part about operating Kafka, it's important to have support for tiered storage.

It would be great if NATS supported tiered storage in a similar fashion (i.e. sealed segments are moved to a different directory, and cached when needed).

**Reply by hpvd** [reply]
Pulsar does also support a handy tiered storage. It can offload a topic to S3 and some others.
for details see: https://pulsar.apache.org/docs/next/tiered-storage-overview/

**Reply by Material-Scientist** [reply]
Since Jetstream stores streams as 8MB blocks, maybe a simple interim solution would just be to have a sidecar that watches for completed blocks, copies them to a remote directory, and then replaces the blocks in the original directory with symlinks that point to the files in the remote directory?

![image](https://github.com/user-attachments/assets/ecda881b-6657-4f64-af29-b79ff9757e70)

Would obviously prefer a native solution by NATS maintainers for prod though.

**Reply by hpvd** [reply]
yep this is interesting.

Or is there an other way:

would it be possible to do not have tiered storage directly, but similar to KV-Store, a new "type" which consists of two components:
- a connection to an external/S3 storage
- a mechanism which offloads (copy and delete /or mark as delete first) a stream depending on a reason to be specified (e.g. message age / TTL, number of included messages, total size...)

**Reply by hpvd** [reply]
looks like this may be related to:
TTL support in KV Store: https://github.com/nats-io/nats-server/issues/3251
S3 compatibility for object store https://github.com/nats-io/nats-server/issues/4871

**Reply by christhegrand** [reply]
We would also love this. We wrote a custom solution to periodically offload data from Jetstream into Google Cloud Storage, but would love to see this functionality built into Jetstream itself.

**Reply by hpvd** [reply]
please see discussion "S3 next level"  for a general approach with further advantage... https://github.com/nats-io/nats-server/discussions/6478
